src/server/pfs/server/driver.go

```diff
@@ -256,46 +251,65 @@ func (d *driver) inspectRepo(txnCtx *txncontext.TransactionContext, repo *pfs.Re
 }
 
 func (d *driver) getPermissions(ctx context.Context, repo *pfs.Repo) ([]auth.Permission, []string, error) {
-	resp, err := d.env.AuthServer.GetPermissions(ctx, &auth.GetPermissionsRequest{
-		Resource: repo.AuthResource(),
-	})
+	resp, err := d.env.AuthServer.GetPermissions(ctx, &auth.GetPermissionsRequest{Resource: repo.AuthResource()})
 	if err != nil {
 		return nil, nil, errors.EnsureStack(err)
 	}
 
 	return resp.Permissions, resp.Roles, nil
 }
 
-func (d *driver) listRepo(ctx context.Context, includeAuth bool, repoType string, cb func(*pfs.RepoInfo) error) error {
+func (d *driver) listRepo(ctx context.Context, includeAuth bool, repoType string, projectsFilter map[string]bool, cb func(*pfs.RepoInfo) error) error {
 	authSeemsActive := true
 	repoInfo := &pfs.RepoInfo{}
 
 	processFunc := func(string) error {
+		// Assume the user meant all projects by not providing any projects to filter on.
+		if len(projectsFilter) > 0 && !projectsFilter[repoInfo.Repo.Project.Name] {
+			return nil
+		}
 		size, err := d.repoSize(ctx, repoInfo.Repo)
 		if err != nil {
 			return err
 		}
 		repoInfo.SizeBytesUpperBound = size
-		if includeAuth && authSeemsActive {
-			permissions, roles, err := d.getPermissions(ctx, repoInfo.Repo)
-			if err == nil {
-				repoInfo.AuthInfo = &pfs.RepoAuthInfo{Permissions: permissions, Roles: roles}
-			} else if auth.IsErrNotActivated(err) {
-				authSeemsActive = false
+
+		// TODO cache the project check
+		// Check if the user can list repo within the project, otherwise check if they can read the repo.
+		if err := d.env.AuthServer.CheckProjectIsAuthorized(ctx, repoInfo.Repo.Project, auth.Permission_PROJECT_LIST_REPO); err != nil {
+			if errors.As(err, &auth.ErrNotAuthorized{}) {
+				if err := d.env.AuthServer.CheckRepoIsAuthorized(ctx, repoInfo.Repo, auth.Permission_REPO_READ); err != nil {
+					if errors.As(err, &auth.ErrNotAuthorized{}) {
+						return nil
+					}
+					return errors.Wrap(err, "could not check user is authorized to access repo")
+				}
+				// Here we know that the user does not have permission to list repos at the project level,
+				// but they should still see the repo because they have read permission.
 			} else {
-				return errors.Wrapf(grpcutil.ScrubGRPC(err), "error getting access level for %q", repoInfo.Repo)
+				return errors.Wrap(err, "could not check user is authorized to access project")
 			}
 		}
+		if authSeemsActive && includeAuth {
+			permissions, roles, err := d.getPermissions(ctx, repoInfo.Repo)
+			if err != nil {
+				if errors.Is(err, auth.ErrNotActivated) {
+					authSeemsActive = false
+					return cb(proto.Clone(repoInfo).(*pfs.RepoInfo))
+				} else {
+					return errors.Wrapf(err, "error getting access level for %q", repoInfo.Repo)
```

Comment:
I'm hitting this error when running opencv as pach:root if I have configured a project other than default as my default project. Do I need to do anything to set up auth for a specific repo? (If this is an extension of the client changes, feel free to ignore)
```
pachctl list repo
rpc error: code = Unknown desc = could not get repos of type "user": ERROR FROM GetByIndex: error getting access level for "NON_DEFAULT-1/images": no role binding exists for REPO NON_DEFAULT-1/images
```
T
he project does exist:
```
pachctl create project NON_DEFAULT-1
projects NON_DEFAULT-1 already exists
```

Reply:
Does `NON_DEFAULT-1/images` repo exist? We create the role binding at repo creation time.

Reply:
It should, at least when I run the example with auth turned off I can see it in `list repo` I tried running the example both before and after activating enterprise auth and it seems to exist without the auth.

Reply:
We figured out the issue. When projects are created in an auth inactive Pachyderm instance, they don't get role bindings created. Once auth is active, auth checks are going to check for the project's role bindings and through an error that they don't existing. However, currently there's a bug where given a repo, we also check the repo's project's role binding, which doesn't exist, but we through an error that is specific to the repo instead of the project.
Fixed here https://github.com/pachyderm/pachyderm/pull/8382/commits/2d360e6cc2fd2b288983bb2365816bc38ee886d2